Produce valid check assertions when dereferencing an invalid pointer symbol

lucasccordeiro opened the pull request:
When dereference failed entirely, we should produce assertions to check for valid objects.

lucasccordeiro committed:
Produce valid check assertions for invalid objects

Once we find an invalid pointer, we should create
assertions before dereferencing the NULL pointer
symbol if we are in READ or WRITE mode.

lucasccordeiro committed:
Added test cases to check for invalid pointer

Added various regression tests to check the assertions for
valid objects.

lucasccordeiro commented:
I'm now running this PR over the SV-COMP benchmarks. The results should be available by tomorrow.

mikhailramalho commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -501,6 +501,8 @@ expr2tc dereferencet::dereference(
     // so that this assignment gets a well typed free value.
     value = make_failed_symbol(type);
+    // Produce assertions to check for invalid objects
+    valid_check(value, guard, mode);
```
I don't think this is the right place to add the valid check, as the comment explains, stuff should break afterward. Does the tests fail if you remove this line?

lucasccordeiro replied:
Yes, the test fails if we remove this line.

lucasccordeiro commented:
This PR increases the number of correct results while keeps the same number of incorrect results.

Master:
```
Statistics:          15201 Files
  correct:            6183
    correct true:     3303
    correct false:    2880
  incorrect:            39
    incorrect true:      6
    incorrect false:    33
  unknown:            8979
  Score:              8766 (max: 24562)
```

This PR:

```
Statistics:          15201 Files
  correct:            6189
    correct true:     3302
    correct false:    2887
  incorrect:            39
    incorrect true:      6
    incorrect false:    33
  unknown:            8973
  Score:              8771 (max: 24562)
```

lucasccordeiro commented:
@mikhailramalho: the method `valid_check` is called only in `build_reference_to`, which is called in https://github.com/esbmc/esbmc/blob/master/src/pointer-analysis/dereference.cpp#L481.

For the regression tests in this PR, the `build_reference_to` method is never called to check for a valid object.

So this comment seems to be deprecated because once we create a failed symbol, we never check whether it is a valid object:

```
    // Dereference failed entirely; various assertions will explode later down
    // the line. To make this a valid formula though, return a failed symbol,
    // so that this assignment gets a well typed free value.
```

Do you have any thoughts?

mikhailramalho commented:
> @mikhailramalho: the method `valid_check` is called only in `build_reference_to`, which is called in https://github.com/esbmc/esbmc/blob/master/src/pointer-analysis/dereference.cpp#L481.
> 
> For the regression tests in this PR, the `build_reference_to` method is never called to check for a valid object.
> 
> So this comment seems to be deprecated because once we create a failed symbol, we never check whether it is a valid object:
> 
> ```
>     // Dereference failed entirely; various assertions will explode later down
>     // the line. To make this a valid formula though, return a failed symbol,
>     // so that this assignment gets a well typed free value.
> ```
> 
> Do you have any thoughts?

What do you mean by deprecated?

The build reference code evaluates the expr recursively so this:
```
    global_var3 = *((uint32_t *)local_var_addr);
```
gets translated into:
```
    global_var3 = *(invalid-pointer);
```

which for me, esbmc gives the following cex using the current master (c3e9fc47e9):
```
Counterexample:

State 1 file foo1.c line 11 function main thread 0
----------------------------------------------------
  local_var_addr = 1 (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001)

State 2 file foo1.c line 12 function main thread 0
----------------------------------------------------
Violated property:
  file foo1.c line 12 function main
  dereference failure: invalid pointer

VERIFICATION FAILED
```

mikhailramalho commented on regression/esbmc/github_382_1/main.c:
```diff
@@ -0,0 +1,8 @@
+  unsigned int *local_var_ptr = (unsigned int *)global_var2;
+  global_var3 = *local_var_ptr;
+  return 0;
+}
```
I don't think this case should fail, we are dereferencing address 0x1, right?

lucasccordeiro replied:
If you compile and run this program, it will fail due to a segmentation fault. We cannot ensure that the address 0x1 can be dereferenced here.

lucasccordeiro committed:
Extended value set to compute an integer constant

Extended the get_value_set method to compute an integer
constant that got turned into a pointer.

fbrausse committed:
[value_set] track values through arithmetic add/sub expressions regardless of type

Prior to this patch, the value set analysis skipped arithmetic expressions that
were not of pointer type, meaning that at least one operand has pointer type.

Now, which operand corresponds to an address is deduced from its value-set not
being empty. While the added regressions github_426_[012] work with this patch,
unfortunately, this approach to determining the pointer operand fails if both
operands have non-empty value sets (ignores the value sets in this case).
This failure case is new since pointer arithmetic syntactically disallowed this
case before.

New known bugs:

regression/esbmc/github_426_3
regression/esbmc/github_426_4

Both perform multiplication and subtraction on uintptr_t. In particular the
failure case mentioned above is triggered by subtracting an address from an
expression already having a value-set.

fbrausse committed:
[value_set] fix for regression/esbmc/github_382_1

fbrausse commented:
I've rebased this to the current master, added 2 commits to track value-sets through add/sub expressions, and added 5 more regression tests (2 of which still fail).

fbrausse committed:
[value_set] fix: assertion was too strong

There are add/sub expressions that themselves are not of pointer-type, but
both operands are.

fbrausse committed:
add regression test: second example from Github issue #382

fbrausse committed:
[value_set] document new parameter + restore old comment in case it is false

rafaelsamenezes reviewed (approved):
LGTM

Just remember to open an issue for those knownbug

fbrausse commented:
> Just remember to open an issue for those knownbug

#92 is a catch all for this.

mikhailramalho commented:
Could please show a before/after this patch on the sv-comp tests? 

GIven that we are changing the memory model, we need to make sure it doesn't break anything else.

shmarovfedor commented:
The results for the SV-COMP benchmarks. Current master:
```
Statistics:          15457 Files
  correct:            7100
    correct true:     4289
    correct false:    2811
  incorrect:            48
    incorrect true:     24
    incorrect false:    24
  unknown:            8309
  Score:             10237 (max: 25029)
```
This PR:
```
Statistics:          15457 Files
  correct:            7069
    correct true:     4272
    correct false:    2797
  incorrect:            47
    incorrect true:     23
    incorrect false:    24
  unknown:            8341
  Score:             10221 (max: 25029)
```

lucasccordeiro commented:
@fbrausse: can I ask you to rebase this branch before we merge it into the master branch?

fbrausse commented:
> can I ask you to rebase this branch before we merge it into the master branch?

Done.

lucasccordeiro commented:
Thanks, @fbrausse.

mikhailramalho commented:
Why we got fewer correct results?

On Wed, 8 Dec 2021, 08:06 Fedor Shmarov, ***@***.***> wrote:

> The results for the SV-COMP benchmarks. Current master:
>
> Statistics:          15457 Files
>   correct:            7100
>     correct true:     4289
>     correct false:    2811
>   incorrect:            48
>     incorrect true:     24
>     incorrect false:    24
>   unknown:            8309
>   Score:             10237 (max: 25029)
>
> This PR:
>
> Statistics:          15457 Files
>   correct:            7069
>     correct true:     4272
>     correct false:    2797
>   incorrect:            47
>     incorrect true:     23
>     incorrect false:    24
>   unknown:            8341
>   Score:             10221 (max: 25029)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/426#issuecomment-988754487>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH2WLGBKZW3DXZJF6MTUP5C3TANCNFSM42OVQZLQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

shmarovfedor commented:
It's hard to say. Probably it's because of timeouts: either there is a slight overhead from this PR, or because of the coarse precision of timeouts (there is always a nondeterministic delay of a few seconds). @fbrausse is analysing the results atm to find out the actual reason.

mikhailramalho commented:
We need to make sure these are actual timeouts and not crashes, since it
reduced both correct true and correct false results, but not the incorrect
ones.

Em qui., 9 de dez. de 2021 às 08:24, Fedor Shmarov ***@***.***>
escreveu:

> It's hard to say. Probably it's because of timeouts: either there is a
> slight overhead from this PR, or because of the coarse precision of
> timeouts (there is always a nondeterministic delay of a few seconds).
> @fbrausse <https://github.com/fbrausse> is analysing the results atm to
> find out the actual reason.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/426#issuecomment-989804544>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH4K7K3M6KAHQUE2F5LUQCNZHANCNFSM42OVQZLQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Mikhail Ramalho.

lucasccordeiro commented:
@fbrausse: given that you're now able to run ESBMC over the SV-COMP benchmarks, can I ask you to revisit this PR?

In particular, we need to know why we have reduced the number of correct true and false. In the first version of this PR, we obtained seven additional correct verification results from the SV-COMP benchmarks (see my initial comments on the top of this PR).

fbrausse commented:
> We need to make sure these are actual timeouts and not crashes, since it
reduced both correct true and correct false results, but not the incorrect
ones.

These are the "diff"-tables created by table-generator from benchexec:

- [unreach-call](https://fbrausse.github.io/esbmc/franz-svcomp/unreach-call/results.2022-01-14_10-17-48.diff.html#/)
- [valid-memsafety](https://fbrausse.github.io/esbmc/franz-svcomp/valid-memsafety/results.2022-01-14_10-19-09.diff.html#/)

(left, 2021-12-07 10:36:59 CET = master; right, 2021-12-03 13:11:34 CET = this PR)

lucasccordeiro commented:
@fbrausse: Can I ask you whether you could investigate the cases where the left column provides a verification outcome while the right column gives an `unknown`?

For these cases, I can see that this PR crashes with these assertions:

```esbmc: ../src/solvers/smt/tuple/smt_tuple_node_ast.h:75: const tuple_node_smt_ast* to_tuple_node_ast(smt_astt): Assertion `ta != nullptr && "Tuple AST mismatch"' failed.```

```Can't construct rvalue reference to array type during dereference (It isn't allowed by C anyway)```

It would also be nice to investigate the case that the master gives incorrect verification result and this PR times out:

```./esbmc-wrapper.py -p ../sv-benchmarks/c/properties/unreach-call.prp -s kinduction --arch 64 ../sv-benchmarks/c/ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--media--usb--ttusb-budget--dvb-ttusb-budget.ko.cil.i```

So, here we could give additional time for this PR to check whether we provide the correct verification result.

fbrausse commented:
Ack. In the first category is also
```
esbmc: ../src/solvers/boolector/boolector_conv.cpp:633: virtual const smt_ast* boolector_convt::mk_ite(smt_astt, smt_astt, smt_astt): Assertion `t->sort->get_data_width() == f->sort->get_data_width()' failed.
```
This looks like it could come from this PR. If so, it might be fixed by https://github.com/esbmc/esbmc/issues/574#issuecomment-1010200413, I'll check.

Regarding
> `Can't construct rvalue reference to array type during dereference (It isn't allowed by C anyway)`

There are a lot (202) of those in the ESBMC-svcomp22 table as well.
